src/NServiceBus.Core/DelayedDelivery/TimeoutManager/TimeoutRecoverabilityBehavior.cs

```diff
@@ -0,0 +1,115 @@
+namespace NServiceBus
+{
+    using NServiceBus.Logging;
+    using NServiceBus.Pipeline;
+    using NServiceBus.Routing;
+    using NServiceBus.Transports;
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    class TimeoutRecoverabilityBehavior : Behavior<ITransportReceiveContext>
+    {
+        public TimeoutRecoverabilityBehavior(string errorQueueAddress, string localAddress, IDispatchMessages dispatcher, BusNotifications notifications, CriticalError criticalError)
+        {
+            this.localAddress = localAddress;
+            this.errorQueueAddress = errorQueueAddress;
+            this.dispatcher = dispatcher;
+            this.notifications = notifications;
+            this.criticalError = criticalError;
+        }
+
+        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
+        {
+            var message = context.Message;
+            var failureInfo = failures.GetFailureInfoForMessage(message.MessageId);
+
+            if (ShouldAttemptAnotherRetry(failureInfo))
+            {
+                try
+                {
+                    await next().ConfigureAwait(false);
+                    return;
+                }
+                catch (Exception exception)
+                {
+                    HandleProcessingFailure(context, message, exception, failureInfo);
+
+                    context.AbortReceiveOperation();
+                    return;
+                }
+            }
+
+            GiveUpForMessage(message, failureInfo);
+
+            await MoveToErrorQueue(context, message, failureInfo);
+        }
+
+        bool ShouldAttemptAnotherRetry(ProcessingFailureInfo failureInfo)
+        {
+            if (failureInfo == null)
+            {
+                return true;
+            }
+            return failureInfo.NumberOfFailedAttempts <= MaxNumberOfFailedRetries;
+        }
+
+        void HandleProcessingFailure(ITransportReceiveContext context, IncomingMessage message, Exception exception, ProcessingFailureInfo failureInfo)
+        {
+            failures.RecordFailureInfoForMessage(message.MessageId, exception);
+
+            Logger.Info($"First Level Retry is going to retry message '{message.MessageId}' because of an exception:", exception);
+
+            var numberOfFailedAttempts = failureInfo?.NumberOfFailedAttempts ?? 0;
+            notifications.Errors.InvokeMessageHasFailedAFirstLevelRetryAttempt(numberOfFailedAttempts, context.Message, exception);
+        }
+
+        void GiveUpForMessage(IncomingMessage message, ProcessingFailureInfo failureInfo)
+        {
+            message.Headers[Headers.FLRetries] = failureInfo.NumberOfFailedAttempts.ToString();
```

Comment:
Can we remove this? It looks the only reason it is set in the first place is for `InvokeMessageHasFailedAFirstLevelRetryAttempt` invocation which has this data already passed via first argument.
@andreasohlund I might missing something here like SC integration maybe? Can you chime in?